sherpa/models/tests/test_model_op.py

```diff
@@ -410,3 +450,245 @@ def test_binop_arithmeticfunction_works(m1, m2):
 
     expected = 2 * np.sin(grid)
     assert y == pytest.approx(expected)
+
+
+class FakeResponse1D:
+    """sherpa.astro.instrument.Response1D requires a PHA. This doesn't.
+
+    This has limited functionality.
+    """
+
+    def __init__(self, arf, rmf):
+        self.arf = arf
+        self.rmf = rmf
+
+    def __call__(self, model):
+        return RSPModelNoPHA(self.arf, self.rmf,
+                             self.arf.exposure * model)
+
+
+class TestBrackets:
+    """Provide a set of model instances for the tests."""
+
+    a = basic.Const1D('a')
+    b = basic.Const1D('b')
+    c = basic.Const1D('c')
+    d = basic.Const1D('d')
+
+    # We don't need to 'load' the model data to use it here
+    tm = basic.TableModel('tm')
+
+    # Convolution-style model (PSF)
+    cm = PSFModel('cm', basic.Const1D('cmflat'))
+
+    # Convolution-style model (PHA)
+    #
+    egrid = np.arange(0.1, 0.5, 0.1)
+    chans = np.arange(1, egrid.size)
+    fake_arf = create_arf(egrid[:-1], egrid[1:], exposure=100.0)
+    fake_rmf = create_delta_rmf(egrid[:-1], egrid[1:])
+
+    arf = ARF1D(fake_arf)
+    rmf = RMF1D(fake_rmf)
+    rsp = FakeResponse1D(fake_arf, fake_rmf)
+
+    # It would be nice to instead use a principled set of states,
+    # but let's just try a somewhat-random set of expressions.
+    #
+    @pytest.mark.parametrize("model,expected",
+                             [(a, "a"),
+                              (abs(a), "abs(a)"),
+                              (abs(a) + b, "abs(a) + b"),
+                              (b + abs(a), "b + abs(a)"),
+                              (abs(a + b), "abs(a + b)"),
+                              (abs(a + b * c), "abs(a + b * c)"),
+                              (abs(a - b * c), "abs(a - b * c)"),
+                              (abs((a + b) * c), "abs((a + b) * c)"),
+                              (abs((a - b) * c), "abs((a - b) * c)"),
+                              (abs((a - b) / c), "abs((a - b) / c)"),
+                              (abs((a * b) - c), "abs(a * b - c)"),
+                              (abs((a / b) - c), "abs(a / b - c)"),
+                              (a * abs(b * (c + d)), "a * abs(b * (c + d))"),
+                              (abs(b * (c + d)) * (a + d), "abs(b * (c + d)) * (a + d)"),
+                              (-a, "-(a)"),
+                              (+a, "+(a)"),
+                              (-a + b, "-(a) + b"),
```

Comment:
Want to try one where you throw in a few extra pairs, e.g. `((-(a) + ((b))))`? Of course, no sensible user will ever do that, but it's be good to have a text to the expected output so we know if it changes later for whatever reason.

Reply:
Done

Reply:
Some tests cover the https://github.com/sherpa/sherpa/pull/1972#discussion_r1545433958 case in more detail, so if I "fix it" we'll see the changes in a few places.

Reply:
Which have now been changed.